llama.cpp

```diff
@@ -1150,6 +1180,17 @@ static bool llama_eval_internal(
 
         lctx.use_buf(ctx0, 0);
 
+        if (lctx.steering_mode != STEERING_OFF && il == lctx.steering_layer) {
```

Comment:
Do I understand it correctly that this is where the tokenized steering prompt gets passed through a layer of the network?

I thought the idea in the post was that the prompt went through all the layers from 0 TO the source layer. As if it were evaluated like a regular prompt.

From what I see here I think it only evaluates it AT the source layer, instead of that layer being the last layer.

Reply:
My C++ is not great, and I am not very familiar yet with how ggml works (yet?) so I'll provide my understanding in pseudocode.

```py
# number of layers for the LLaMa-7B model
llama_layers_7B = 32

source_layer = 2
steering_layer = 20
steering_prompt_add = "Love"
steering_prompt_sub = "Hate"

llama_prompt = "I love "

tokens_add = tokenize(steering_prompt_add)
tokens_sub = tokenize(steering_prompt_sub)


def evaluate_prompt_until(tokens, stop_layer):
    result = ggml_init(tokens) # not sure what to call it, but it's certainly not a string. Is it a tensor?
    # iteratively pass the initial tokens through the layers, but stop at the specified stop_layer
    for il in range(0, stop_layer):
        result = ggml_pass_through_layer(result, il)
    return result

steer_vector_add = evaluate_prompt_until(tokens_add, source_layer)
steer_vector_sub = evaluate_prompt_until(tokens_sub, source_layer)

steer_vector = steer_vector_add - steer_vector_sub

tokens_llama_prompt = tokenize(llama_prompt)
# this function evaluates the prompt like llama.cpp has always done, but adds the steering vector
# at layer=steer_at_layer
output = ggml_evaluate_with_steering(tokens_llama_prompt, steer_at_layer=steering_layer, steer_vector=steer_vector)
print(unembed(output))
```

I'm not sure my understanding is 100% correct, but this is what I took away from the paper. Once again, in English:

1. For the steer_add prompt, compute the model up until layer=source layer, cache result as steer_add_vector
2. Do the same for steer_sub
3. Add the steer_add_vector and steer_sub_vector vectors. Save as `steer_vector`
4. Evaluate prompt as usual, except add the steer_vector to the input for layer=steer_layer
5. That's it!

Reply:
That's very close to how I did it.

Except I have only one vector and the coefficient is actually used in a general way in the llama model code.

In more Python-like pseudocode:

```py
source_layer = 2
target_layer = 20
steering_coefficient = 4
steering_prompt_add = "Love"
steering_prompt_sub = "Hate"

# a prompt that generally emits negative sentiment
llama_prompt = "People who commit crimes are "

tokens_sub = tokenize(steering_prompt_sub)
tokens_add = tokenize(steering_prompt_add)

# this is part of the llama core
def llama_evaluate(tokens, st_vector=None, st_coefficient=0, st_mode='OFF', st_layer):
    
    inpL = model.tok_embeddings[tokens, :] # for every token there is a 4096-element vector
    
    for il in range(0, model.n_layer):
        
        if st_mode != 'OFF' and il == st_layer:
            if st_mode == 'WRITE':
                st_vector = st_vector + inpL*st_coefficient
                break # we can stop processing the layers to save time
            else:
                inpL = inpL + st_vector*st_coefficient
        
        cur = norm(inpL)
        cur = attention(cur)
        cur = feed_fwd(cur - inpL)
        
        inpL = cur
    
    logits = model.output % inpL
    st_vector.assign(st_vector) # copy back the result
    return logits

steering_vector = []
evaluate_prompt_until(tokens_add, steering_vector, +1, 'WRITE', source_layer)
evaluate_prompt_until(tokens_sub, steering_vector, -1, 'WRITE', source_layer)

tokens_llama_prompt = tokenize(llama_prompt)
# this function evaluates the prompt like llama.cpp has always done, but adds the steering vector
# at layer=steer_at_layer
output = llama_evaluate(tokens_llama_prompt, steering_vector, steering_coefficient, 'READ', target_layer)
print(unembed(output))
```

Reply:
Actually, the code could also be simplified, there is already functionality to export the vectors after the last layer.

Reply:
> Do I understand it correctly that this is where the tokenized steering prompt gets passed through a layer of the network?
>```c
>if (lctx.steering_mode != STEERING_OFF && il == lctx.steering_layer) {
>```

I didn't answer the question, did I?

This part is ony for the steering vector, either to add the embeddings to it or to add it back the the embeddings.

When creating the steering vectors in the first place, it is also the last layer because of the `break,` this is just for performance to not compute the later layers, the output of which we don't need.